### PR TITLE
mosquitto: 1.6.12 -> 2.0.10

### PR DESCRIPTION
**Pull request by prusnak**
###### Motivation for this change

* new major 2.x version: https://mosquitto.org/blog/2020/12/version-2-0-0-released/
* see https://mosquitto.org/documentation/migrating-to-2-0/

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by prusnak** [comment]
@GrahamcOfBorg build mosquitto mosquitto-1 mosquitto-2

File: pkgs/servers/mqtt/mosquitto/2.x.nix [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+, cmake
+, docbook_xsl
+, libxslt
+, c-ares
+, libuuid
+, libuv
+, libwebsockets_3_1
+, openssl
+, systemd ? null
+, withSystemd ? stdenv.isLinux
```
Comment (SuperSandro2000) [inline pkgs/servers/mqtt/mosquitto/2.x.nix:13]:
```suggestion
, withSystemd ? stdenv.isLinux
, systemd
```

File: pkgs/servers/mqtt/mosquitto/1.x.nix [hunk 2]
```diff
@@ -1,15 +1,26 @@
...
+, cmake
+, docbook_xsl
+, libxslt
+, c-ares
+, libuuid
+, libuv
+, libwebsockets_3_1
+, openssl
+, systemd ? null
+, withSystemd ? stdenv.isLinux
```
Comment (SuperSandro2000) [inline pkgs/servers/mqtt/mosquitto/1.x.nix:13]:
```suggestion
, withSystemd ? stdenv.isLinux
, systemd
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 110863` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
5 packages built:
</summary>
<ul>
<li>collectd</li>
<li>collectd-data</li>
<li>domoticz</li>
<li>mosquitto (mosquitto-1)</li>
<li>mosquitto-2</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

mosquitto-2:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern '/sbin/ldconfig' doesn't match anything in file 'lib/CMakeLists.txt'
substituteStream(): WARNING: pattern '/sbin/ldconfig' doesn't match anything in file 'lib/cpp/CMakeLists.txt'
substituteStream(): WARNING: pattern '/sbin/ldconfig' doesn't match anything in file 'src/CMakeLists.txt'
```

Please check the offending substituteInPlace for typos or changes in source.
</details>
mosquitto:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern '/sbin/ldconfig' doesn't match anything in file 'lib/CMakeLists.txt'
substituteStream(): WARNING: pattern '/sbin/ldconfig' doesn't match anything in file 'lib/cpp/CMakeLists.txt'
substituteStream(): WARNING: pattern '/sbin/ldconfig' doesn't match anything in file 'src/CMakeLists.txt'
```

Please check the offending substituteInPlace for typos or changes in source.
</details>

File: pkgs/servers/mqtt/mosquitto/1.x.nix (hunk 2 above)
Comment (prusnak) [reply pkgs/servers/mqtt/mosquitto/1.x.nix]: In 8c00b51

File: pkgs/servers/mqtt/mosquitto/2.x.nix (hunk 1 above)
Comment (prusnak) [reply pkgs/servers/mqtt/mosquitto/2.x.nix]: In 9a46e5c

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 110863` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages marked as broken and skipped:
</summary>
<ul>
<li>collectd</li>
<li>collectd-data</li>
</ul>
</details>
<details>

  <summary>1 package failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>domoticz: log <a href="https://termbin.com/w3wm">https://termbin.com/w3wm</a></li>
  </ul>
</details>
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>mosquitto (mosquitto-1)</li>
<li>mosquitto-2</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 110863` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
5 packages built:
</summary>
<ul>
<li>collectd</li>
<li>collectd-data</li>
<li>domoticz</li>
<li>mosquitto (mosquitto-1)</li>
<li>mosquitto-2</li>
</ul>
</details>

**Comment by fabaff** [comment]
Just tow remarks:

- cJSON is a [new requirement](https://github.com/eclipse/mosquitto#build-dependencies) -> https://github.com/NixOS/nixpkgs/pull/107169
- 2.0.6 was [release](https://mosquitto.org/blog/2021/01/version-2-0-6-released/)

**Comment by peterhoeg** [comment]
I don't really think we need to keep both v1 and v2. The changes are relatively minor and therefore quite easy to work around in the nixos module.

**Comment by prusnak** [comment]
> I don't really think we need to keep both v1 and v2. The changes are relatively minor and therefore quite easy to work around in the nixos module.

Reworked the PR as a simple version update. Added cJSON to buildInputs.

@peterhoeg I haven't touched the nixos module, will you update `nixos/modules/services/networking/mosquitto.nix`?

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 110863` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages marked as broken and skipped:
</summary>
<ul>
<li>collectd</li>
<li>collectd-data</li>
</ul>
</details>
<details>

  <summary>1 package failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>domoticz: log <a href="https://termbin.com/1k27">https://termbin.com/1k27</a></li>
  </ul>
</details>
<details>
<summary>
1 package built:
</summary>
<ul>
<li>mosquitto</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 110863` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
4 packages built:
</summary>
<ul>
<li>collectd</li>
<li>collectd-data</li>
<li>domoticz</li>
<li>mosquitto</li>
</ul>
</details>

**Comment by prusnak** [comment]
@GrahamcOfBorg build mosquitto

**Comment by prusnak** [comment]
@GrahamcOfBorg test mosquitto

**Comment by mweinelt** [comment]
Authentication with Mosquitto broke (`Connection refused: not authorized`) in the nixos home-assistant test due to this upgrade. This is our setup:

https://github.com/NixOS/nixpkgs/blob/a05293a93c8498ee3b2e72afc91cbfbd04547918/nixos/tests/home-assistant.nix#L15-L23

The credentials are being reused for `home-assistant` and `mosquitto_sub`:

https://github.com/NixOS/nixpkgs/blob/a05293a93c8498ee3b2e72afc91cbfbd04547918/nixos/tests/home-assistant.nix#L38-L42

https://github.com/NixOS/nixpkgs/blob/a05293a93c8498ee3b2e72afc91cbfbd04547918/nixos/tests/home-assistant.nix#L79-L86

Looking for advice.

**Comment by peterhoeg** [comment]
I wanted to, but didn't have a chance to work through the module, so I suggest we do this.

1. revert this 
2. introduce a new mosquitto2 with this version
3. introduce mosquitto1 with the latest 1.6.x
4. make mosquitto an alias for mosquitto1

1.6 is still maintained (and there are some security fixes we need) so we can then always add proper support in the nixos module.

**Comment by mweinelt** [comment]
```
hass # [    4.109957] mosquitto[761]: 1617846007: The 'bind_address' option is now deprecated and will be removed in a future version. The behaviour will default to true.
hass # [    4.110884] mosquitto[761]: 1617846007: The 'port' option is now deprecated and will be removed in a future version. Please use 'listener' instead.
hass # [    4.112071] mosquitto[761]: 1617846007: mosquitto version 2.0.10 starting
hass # [    4.112542] mosquitto[761]: 1617846007: Config loaded from /nix/store/mkrjj35ij3hld114h3gz4x2cmkiyxj2d-mosquitto.conf.
hass # [    4.114645] mosquitto[761]: 1617846007: Warning: ACL pattern '#' does not contain '%c' or '%u'.
hass # [    4.117063] mosquitto[761]: 1617846007: Opening ipv4 listen socket on port 1883.
hass # [    4.117545] mosquitto[761]: 1617846007: mosquitto version 2.0.10 running
```

Two deprecation warnings, which should probably have been addressed in this pull request. @prusnak @SuperSandro2000 

And probably my issue: A warning for the ACL pattern `topic readwrite #`. Great.

**Comment by mweinelt** [comment]
image

So one problem was that `checkPasswords` was not set, so the password file was not included.

**Comment by mweinelt** [comment]
Second issue is, that the following command exits with 1 ("Timed out") since 2.0.

> "mosquitto_sub -V mqttv311 -t home-assistant/test -u ${mqttUsername} -P '${mqttPassword}' -W 1 -t '*'"

Also not sure why `-t` was given twice, makes no sense to me.

**Pull request merged**